### ISSUE TITLE
Flutter doesn't respect Android's `android:usesCleartextTraffic="false"` settings

### ISSUE THREAD

**Issue by maRci002** [flutter/flutter#106678]
## Steps to Reproduce

1. disable clear text for both `release` and `debug` builds
    * **release:** `android/app/src/main/AndroidManifest.xml` add `android:usesCleartextTraffic="false"` to `<application>` element's attributes
    * **debug:** `android/app/src/debug/AndroidManifest.xml` add `<application android:usesCleartextTraffic="false" />` element for the `<manifest>` element
2. make some `http` request
3. no error is raised no matter if I build against `debug` or `release` mode

## Related Pull Requests

* `android:usesCleartextTraffic="false"` should be respected since flutter/engine/pull/20218
   https://github.com/flutter/engine/commit/dd2c5a1b8bb4fbd76b6882d2daa00a6dc26c14ca#diff-461938e51a0b15219735ea1fdca7abc739f390536deb8239a3598483fa577438R145-R149
* flutter/engine#25299 removed this feature somewhy
   https://github.com/flutter/engine/commit/f0476334b3227f24efcf7636a3ac6086a1576d8c#diff-2777e9e1be51ba3b8b43d3e74741a668427e4eb3967ed759772a9dbe81880605L240-L242
   I think on iOS the result is the same which also ignores the config according to these changes:
   ```dart
   settings.may_insecurely_connect_to_all_domains = true;
   settings.domain_network_policy = "";
   ```
* flutter/engine#25940 fully removed unused Java variables
   https://github.com/flutter/engine/pull/25940/files#diff-461938e51a0b15219735ea1fdca7abc739f390536deb8239a3598483fa577438L147-L165

   > inside `ApplicationInfoLoader.java` which results in calling the underlying xml parsing via the NetworkSecurityPolicy. This results in additional I/O contention in the main thread.

   According to doc: https://developer.android.com/guide/topics/manifest/application-element#usesCleartextTraffic it can also obtained by [ApplicationInfo.flags](https://developer.android.com/reference/android/content/pm/ApplicationInfo#flags) but this may not respect [Network security configuration](https://developer.android.com/training/articles/security-config) file.

## Code sample

**android/app/src/main/AndroidManifest.xml**
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.disable_clear_text">
   <uses-permission android:name="android.permission.INTERNET"/> <!--Make sure Internet is avaible for release builds-->

  <!--Add android:usesCleartextTraffic="false"-->
   <application
        android:label="disable_clear_text"
        android:name="${applicationName}"
        android:usesCleartextTraffic="false"
        android:icon="@mipmap/ic_launcher">
```

**android/app/src/main/AndroidManifest.xml**
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.disable_clear_text">
    <uses-permission android:name="android.permission.INTERNET"/>
    <application android:usesCleartextTraffic="false" /> <!--Add this line and make sure it is set to false-->
</manifest>
```

```dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Disable Clear Text',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final client = HttpClient();
  var _httpResult = 'Press "Fetch button"';

  void _fetchClearText() async {
    try {
      final url = Uri.parse('http://example.com');
      final request = await client.getUrl(url);
      final response = await request.close();
      final responseData = await response.transform(utf8.decoder).join();

      _httpResult = 'type: ${response.runtimeType}\ncontentLength: ${response.contentLength}\ncontainsHtmlTag: ${responseData.contains('<html>')}';
    } catch (e) {
      _httpResult = e.toString();
    }

    setState(() {});
  }

  @override
  void dispose() {
    client.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(_httpResult),
              ElevatedButton(
                onPressed: _fetchClearText,
                child: const Text('Fetch'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

Optionally override `MainActivity.java` 's consturctor just to make sure whether clear text traffic is permitted (search for `isCleartextTrafficPermitted` inside logcat which should yield: `com.example.disable_clear_text.MainActivity: isCleartextTrafficPermitted: false`)
```java
package com.example.disable_clear_text;

import android.os.Build;
import android.security.NetworkSecurityPolicy;
import android.util.Log;

import io.flutter.embedding.android.FlutterActivity;

public class MainActivity extends FlutterActivity {
    public MainActivity() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            boolean isCleartextTrafficPermitted = NetworkSecurityPolicy.getInstance().isCleartextTrafficPermitted();
            Log.w(MainActivity.class.toString(), "isCleartextTrafficPermitted: " + isCleartextTrafficPermitted);
        }
    }
}
```

[Here is the minimal repository](https://github.com/maRci002/disable_clear_text) to reproduce this issue and [here is the specific commit which contains the above code snippets](https://github.com/maRci002/disable_clear_text/commit/fcb8eeeb3e7b15532b18765822cfa7e8c87de17c).

<details>
  <summary>flutter doctor -v</summary>

```
PS C:\Users\maRci002> flutter doctor -v
[√] Flutter (Channel stable, 3.0.3, on Microsoft Windows [Version 10.0.19044.1766], locale hu-HU)
    • Flutter version 3.0.3 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 676cefaaff (5 days ago), 2022-06-22 11:34:49 -0700
    • Engine revision ffe7b86a1e
    • Dart version 2.17.5
    • DevTools version 2.12.2

[√] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at C:\Users\maRci002\AppData\Local\Android\sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code (version 1.68.1)
    • VS Code at C:\Users\maRci002\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.42.0

[√] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 12 (API 31) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.19044.1766]
    • Chrome (web)                 • chrome        • web-javascript • Google Chrome 102.0.5005.115
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 103.0.1264.37

[√] HTTP Host Availability
    • All required HTTP hosts are available
```

</details>

**Comment by maheshmnj** [comment]
Hi @maRci002, Thanks for filing the issue. I am able to reproduce the issue on the latest stable and master channel. unsecured network requests still work with `android:usesCleartextTraffic= false`

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 3.0.3, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.3 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 676cefaaff (5 days ago), 2022-06-22 11:34:49 -0700
    • Engine revision ffe7b86a1e
    • Dart version 2.17.5
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.43.20220621

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.53

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1376, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.1.0-0.0.pre.1376 on channel master at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3598f20002 (9 hours ago), 2022-06-26 18:19:04 -0400
    • Engine revision ee56813c14
    • Dart version 2.18.0 (build 2.18.0-228.0.dev)
    • DevTools version 2.14.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13C100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.43.20220621

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.53

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by GaryQian** [comment]
@xster It seems that the fix for this is to reland the reverted code in https://github.com/flutter/engine/pull/25299

Can you give context on why this was reverted and if it is safe to reland?

**Comment by maRci002** [comment]
This is weird for me:
* On Android [Network security configuration](https://developer.android.com/training/articles/security-config) is still [consumed](https://github.com/flutter/engine/blob/0cc828c9b611316ee18b96e01b29c23ba126dd56/shell/platform/android/io/flutter/embedding/engine/loader/ApplicationInfoLoader.java#L154)
  * I just need to tell Flutter where is my [configuration](https://github.com/flutter/engine/blob/0cc828c9b611316ee18b96e01b29c23ba126dd56/shell/platform/android/io/flutter/embedding/engine/loader/ApplicationInfoLoader.java#L67)
  * I created a branch https://github.com/maRci002/disable_clear_text/tree/feat/network_security_config and this [commit](https://github.com/maRci002/disable_clear_text/commit/ce59a62493ca362258db2822d50f2ccbe5dc7397) tries to use Network security configuration
* On iOS domain [network policy is just constant empty](https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L157)

**Comment by kevinbarja** [comment]
after 5 months, something new guys?

**Comment by maRci002** [comment]
I think This is why this feature was removed: https://docs.flutter.dev/release/breaking-changes/network-policy-ios-androidt

**Comment by Nobler** [comment]
> This is weird for me:
> 
> * On Android [Network security configuration](https://developer.android.com/training/articles/security-config) is still [consumed](https://github.com/flutter/engine/blob/0cc828c9b611316ee18b96e01b29c23ba126dd56/shell/platform/android/io/flutter/embedding/engine/loader/ApplicationInfoLoader.java#L154)
>   
>   * I just need to tell Flutter where is my [configuration](https://github.com/flutter/engine/blob/0cc828c9b611316ee18b96e01b29c23ba126dd56/shell/platform/android/io/flutter/embedding/engine/loader/ApplicationInfoLoader.java#L67)
>   * I created a branch https://github.com/maRci002/disable_clear_text/tree/feat/network_security_config and this [commit](https://github.com/maRci002/disable_clear_text/commit/ce59a62493ca362258db2822d50f2ccbe5dc7397) tries to use Network security configuration
> * On iOS domain [network policy is just constant empty](https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L157)

`<base-config>` is not being handled, see https://github.com/flutter/flutter/issues/65841#issuecomment-692810524.
After my testing, even `<domain-config>` is not hanled as well, I guess the engine does not consume the arg. I don't know so much about the c++ side, just sure about the java side actually handled:
https://github.com/flutter/engine/blob/0cc828c9b611316ee18b96e01b29c23ba126dd56/shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java#L286

**Comment by brianquinlan** [comment]
@maheshmnj If I understand this issue correctly, it should not have the "dependency: dart" or "dependency: dart-io" labels because Flutter can decide what URLs are loadable.

See https://github.com/flutter/flutter/issues/95341#issuecomment-1799763864

**Comment by slaci** [comment]
I can confirm the same unexpected behavior. As I found out, for the [http](https://pub.dev/packages/http) lib the cupertino and/or cronet clients must be used (they are mentioned in the readme) which respect the native constraints.

When using [dio](https://pub.dev/packages/dio), then the [native_dio_adapter](https://pub.dev/packages/native_dio_adapter) package can be used which uses the above mentioned clients under the hood.

When using the default IO client, then I guess developers must always implement this http blocking logic somehow (not entirely clear how) as it is "enabled" by default. Normally cleartext is only enabled on Android below 8.1 if you don't put the `android:usesCleartextTraffic="false"` config into the manifest, as that value was `true` by default in those early versions.

**Comment by E2-Veera** [comment]
i've the same problem. I'm confused in the [flutter documentation](https://docs.flutter.dev/release/breaking-changes/network-policy-ios-android) says http insecure connection will be blocked by default. but still working for me, Even after adding this `android:usesCleartextTraffic="false"` and in iOS also the same behaviour is identified.

More details : [link](https://stackoverflow.com/questions/79222050/http-api-call-works-despite-cleartexttrafficpermitted-false-in-android-app-wit)

[✓] Flutter (Channel stable, 3.22.3, on macOS 14.3.1 23D60 darwin-arm64, locale en-IN)
    • Flutter version 3.22.3 on channel stable at /Users/veerasivaraman/Documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b0850beeb2 (4 months ago), 2024-07-16 21:43:41 -0700
    • Engine revision 235db911ba
    • Dart version 3.4.4
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/veerasivaraman/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] VS Code (version 1.84.2)
    • VS Code at /Users/veerasivaraman/Downloads/Visual Studio Code.app/Contents
    • Flutter extension version 3.100.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.3.1 23D60 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 131.0.6778.86
  
[✓] Network resources
    • All expected network resources are available.

• No issues found!

Any fix for this ??

**Comment by m-rm** [comment]
> i've the same problem. I'm confused in the [flutter documentation](https://docs.flutter.dev/release/breaking-changes/network-policy-ios-android) says http insecure connection will be blocked by default. but still working for me, Even after adding this `android:usesCleartextTraffic="false"` and in iOS also the same behaviour is identified.

The first link you posted contains the following information:

> Important
> 
> This change over-restricted HTTP access on local networks beyond the restrictions imposed by mobile platforms ([flutter/flutter#72723](https://github.com/flutter/flutter/issues/72723)).
> 
> This change has since been reverted.

It would still be nice to have a HTTPS-only option when local network access is not required.

**Comment by jmagman** [comment]
From https://developer.android.com/about/versions/17/behavior-changes-all#uses-clear-traffic-deprecation:

> In a future release, we plan to deprecate the usesCleartextTraffic element. Apps that need to make unencrypted (HTTP) connections should migrate to using a [network security configuration](https://developer.android.com/privacy-and-security/security-config) file, which lets you specify which domains your app needs to make cleartext connections to.

See also https://github.com/flutter/flutter/issues/106678.

Given that, we're unlikely to do any work to make deprecated `usesCleartextTraffic` work than it does now. Closing.
